src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Internal.hs

```diff
@@ -179,20 +196,22 @@ toField :: Arithmetic a => a -> VarField
 toField = toZp . fromConstant . fromBinary @Natural . castBits . binaryExpansion
 
 -- TODO: Remove the hardcoded constant.
-toVar :: Arithmetic a => [Natural] -> Constraint a -> Natural
+toVar :: Arithmetic a => [Var i] -> Constraint a i -> Natural
 toVar srcs c = force $ fromZp ex
     where
         r  = toZp 903489679376934896793395274328947923579382759823 :: VarField
         g  = toZp 89175291725091202781479751781509570912743212325 :: VarField
-        v  = (+ r) . fromConstant
+        varF (NewVar w) = w
+        varF (InVar _)  = 0
```

Comment:
This one might turn out to be problematic if the only variable we depend on is an input variable.

All in all, I think we should change our value generation algorithm to generate it from witness instead: the basic idea is to view witness as a syntax tree which we can Merkle-hash.

P.S. Then we would need `Rep i` to be hashable; this in turn means that we might want to migrate to hashmaps in arithmetic circuit to reduce the number of constraints we impose on `i`, maybe?

Reply:
Yes, we could have multiple constraints with the same constants that depend on different subsets of input variables. This would give the same new variable every time if I am not mistaken. So, it must be fixed in this PR.

Reply:
As a hotfix, we can require `ToConstant (Rep i) Natural` before we migrate to the proper solution